### PR TITLE
Add model of iCub and fix instability of grocery-ground

### PR DESCRIPTION
**Pull request by Jialn**
Add the model of open source humanoid icub to social robot. http://www.icub.org/
image

Also added a standing and walking env based on icub , as well as a training example.
It's not solved yet. The training example can only make the robot standing longer for now.

### CODE REVIEW

File: examples/train_icub_walk_sac.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""Train and Eval ICub.
+
+A simple test for SocialBot-ICub env use TF-Agents
+The original file is from the PPO example of TF-Agents:
+    https://github.com/tensorflow/agents
+Some parameters are modifed suite the environment.
```
[the remark this answers is not in the document]
Comment (Jialn) [reply examples/train_icub_walk_sac.py]: The original tf-agent SAC does not support Parallel Environment. gin configuration won't do. Will change to gin configuration latter when alf SAC merged and tested on this environment.

File: python/social_bot/models/icub/l_hand/model.config [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    <version>1.0</version>
+    <sdf version='1.6'>l_hand.sdf</sdf>
+    
+    <author>
+        <name>Marco Randazzo</name>
+        <email>marco.randazzo@iit.it</email>
+    </author>
+    
+    <description>
+        Model for the hand of iCub humanoid robot. For more information check icub.org.
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/models/icub/l_hand/model.config]: Added

**Comment by emailweixu** [comment]
Can you also document the changes to the iCub model compared to the original model?

**Comment by Jialn** [comment]
> Can you also document the changes to the iCub model compared to the original model?

Sure

**Comment by Jialn** [comment]
commit cd1ccd3 @ 1M steps
f5guu-4q92u

**Comment by Jialn** [comment]
commit 7789a00 @10M steps
4fm5e-rhs28

**Comment by emailweixu** [comment]
Does the recent commit solve the stability problem?

**Comment by Jialn** [comment]
> Does the recent commit solve the stability problem?
Part of. Like possible crash when inserting agent pr2 and icub solved, and we don’t need to sleep for a while after inserting operation now. Still possible to get stuck into original pos.

**Comment by Jialn** [comment]
A possible reason is that all the inserted objects after the world has been created has a default pos of (0,0,0). So the objects are reset to (0,0,0) after self._world.reset() and before 'random_move_obj'. That caused lots of unnatural collisions. 
Gazebo implements a standalone 'ContactManager' for collision processing rather than use the one from the phy-engine ode or bullet. Those unnatural collisions together with a higher action and larger sim time step might lead to super large number somewhere during the collision computing progress.

**Comment by emailweixu** [comment]
What if we don't use world.reset()?

**Comment by emailweixu** [comment]
Actually, I cannot find "self._world.reset()" in the code. What I found is "self._agent.reset()". So perhaps you can use a different way of reset agent state so that it won't collide with other objects after reset.

**Comment by Jialn** [comment]
Did some experiments for the ration reward components, record the result here. That really matters a lot. 

ICubWalkCtrlCostx1.0

        reward = 1.0 + 6.0 * min(walk_vel, 1.0) - 1.0 * ctrl_cost

ICubWalkCtrlCostx2.0

        reward = 1.0 + 6.0 * min(walk_vel, 1.0) - 2.0 * ctrl_cost

ICubWalkJontPosx1.0

        pose_ctrl_cost = np.sum(np.abs(self.joint_pos))/self.joint_pos.shape[0]
        ctrl_cost = action_ctrl_cost + pose_ctrl_cost
        reward = 1.0 + 6.0 * min(walk_vel, 1.0) - 1.0 * ctrl_cost

ICubWalkJontPosx0.5

        ctrl_cost = action_ctrl_cost + 0.5 * pose_ctrl_cost
        reward = 1.0 + 6.0 * min(walk_vel, 1.0) - 1.0 * ctrl_cost


ICubWalkJointPosx0.5Chest 

        The same as ICubWalkJontPosx0.5 except that one pose in observation use chest link instead of head link. 


ICubWalkJointPosx0.5Chest_higherLR 

        A higher learning rate (5e-4) version of ICubWalkJointPosx0.5Chest 


image

**Comment by Jialn** [comment]
> Actually, I cannot find "self._world.reset()" in the code. What I found is "self._agent.reset()". So perhaps you can use a different way of reset agent state so that it won't collide with other objects after reset.

Yes, there are "self._world.reset()" for now. I found "self._agent.reset()" can not reset the agent to initial state. I don't know why, seems only pose of root link is reset. The joints are not reset and become a mess, might be caused by the super fast movements of root link. I guess that's why only pr2 and icub have the problem, model only has wheels won't be affected). 
So I was testing on my local branch, added "self._world.reset()" which reset everything, sovled the initial state problem, as well as stucking into original map point problem. But brings the collisions. 

Adding "self._world.reset()" and then immediately moving the objects without stepping the word seems to be a way out. Everything looks fine now. I'll do more testing.

**Comment by Jialn** [comment]
The reset procedure does solve some instability problem during reset. But pr2 and icub_with_hands can still has a change to fail during the simulation. Some tests has shown that the instability comes from super high velocity during a sim time steps, which caused by the interpenetration correction or high effort limit/ low inertial mass. 

One problem is interpenetration correction. It applies an impulse that pushes the object away from interpenetrating collision, but if the impulse is too large it causes the object to "bounce/fly away". Those unrealistically high velocity / momentum bring numerical instability.

Another reason is, some complex model like icub_with_hands and pr2 have fingers or grippers which are rather slight in mass. If a relatively higher action value is given, numerical instability happens.
For example, each finger link of icub_with_hands has a mass of 0.01kg and effort limit 10N*m (1000 N*cm) by default, with about 1cm inertial center away from the joint(moment arm), there would be ~ 10,000m/s^2 acceleration (1000/(1*0.01)), such a huge acceleration might also lead to unrealistically high velocity or distance(twice integral of acceleration) in several simulator time-step, and cause numerical instability during the computation of collision.

So, changes to solve the instability problem include:

- Insert the agent to world file before creating the world rather than insert it dynamically;
- Use world.reset() rather than agent.reset() in gro2env;
- Set sim_time_step to 1ms. Although pr2 and icub(without hands) can work with a 2.5ms sim_time_step, we will still use 1ms as gripper of pr2 also require 1ms sim precision)
- Set the parameters contact_max_correcting_vel for ode engine in world file, which is used to for interpenetration correction; I'm quite sure if this term is useful.
- Increase mass of the icub finger to 0.1 (That's important, I've tried 0.05 or 0.025, the model still has a chance to fail with 1ms sim_time_step);
- Scale the finger's collision by 0.9 to remove possible interpenetration. 
- Set the limit of effort for fingers to 0.5N*m (~5Kg*cm). That's 20x smaller compared to the default 10M*m, but it's enough, consider length of icub finger;

**Comment by Jialn** [comment]
It's ready to merge now. Please review again.

File: python/social_bot/envs/icub_walk.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        ]).astype(np.float32)
+        obs = np.concatenate(
+            (agent_poses, agent_vel, joint_pos, joint_vel, foot_contacts),
+            axis=0)
+        return obs
+
+    def step(self, action):
+        """
+        Args:
+            action (float):  action is a vector whose dimention is
```
Comment (emailweixu) [inline python/social_bot/envs/icub_walk.py]: action (np.array)

File: python/social_bot/envs/grocery_ground.py [hunk 4]
```diff
@@ -53,9 +53,9 @@ def __init__(self, reward_shaping=False, random_goal=False, **kwargs):
         super(GroceryGroundGoalTask, self).__init__(**kwargs)
         self._reward_shaping = reward_shaping
         self._random_goal = random_goal
-        self._goal_name = 'cube_20k'
+        self._goal_name = 'cafe_table'
         self._object_list = [
-            'coke_can', 'table', 'bookshelf', 'cube_20k', 'car_wheel',
+            'coke_can', 'table', 'bookshelf', 'car_wheel',
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: I don't know why but it seems the surface of cube20k could cause instability. Sometimes it bounces the agent or the moved goal to fly away. Similar problem had been found with trashcan, which is removed before.

File: python/social_bot/envs/grocery_ground.py [hunk 5]
```diff
@@ -53,12 +53,16 @@ def __init__(self, reward_shaping=False, random_goal=False, **kwargs):
         super(GroceryGroundGoalTask, self).__init__(**kwargs)
         self._reward_shaping = reward_shaping
         self._random_goal = random_goal
-        self._goal_name = 'cube_20k'
+        self._goal_name = 'coke_can_on_table'
+        self._static_object_list = [
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: static_object_list contains the fixed model in the world file(fixed table, walls, etc.) , which should be added to the vocabulary.

File: pygazebo/pygazebo.cc [hunk 6]
```diff
@@ -290,9 +290,10 @@ class Agent : public Model {
                         double i,
                         double d,
                         double max_force) {
-    auto pid = gazebo::common::PID(p, i, d);
+    auto pid = gazebo::common::PID(p*max_force, i*max_force, d*max_force);
```
[the remark this answers is not in the document]
Comment (Jialn) [reply pygazebo/pygazebo.cc]: OK

File: pygazebo/pygazebo.cc [hunk 7]
```diff
@@ -290,9 +290,10 @@ class Agent : public Model {
...
     pid.SetCmdMax(max_force);
     pid.SetCmdMin(-max_force);
+    pid.SetIMax(max_force * 0.1); // limit i term
```
[the remark this answers is not in the document]
Comment (Jialn) [reply pygazebo/pygazebo.cc]: OK

File: python/social_bot/models/agent_cfg.json [hunk 8]
```diff
@@ -1,22 +1,84 @@
...
+            "iCub::l_shoulder_pitch",
+            "iCub::l_shoulder_roll",
+            "iCub::l_shoulder_yaw",
+            "iCub::l_elbow",
+            "iCub::neck_pitch",
+            "iCub::neck_roll",
+            "iCub::r_shoulder_pitch",
+            "iCub::r_shoulder_roll",
+            "iCub::r_shoulder_yaw",
+            "iCub::r_elbow"],
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/models/agent_cfg.json]: not for now. we could add them later if a future task would need these joints.

File: examples/train_grocery_ground.py [hunk 9]
```diff
@@ -17,15 +17,14 @@
 A simple test for SocialBot-GroceryGround env use TF-Agents
 The original file is from the PPO example of TF-Agents:
     https://github.com/tensorflow/agents
-Some parameters are modifed suite the environment.
+Some parameters are modified to suite the environment.
```
[the remark this answers is not in the document]
Comment (Jialn) [reply examples/train_grocery_ground.py]: Sure

File: python/social_bot/envs/icub_walk.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+class ICubWalk(GazeboEnvBase):
+    """
+    The goal of this task is to make the agent icub standing and walking.
+    Joints are controllable with force, or with target velocity if the internal
+        pid controller is used.
+    Observation is a single ndarray with internal joint states, and agent pose.
+    Reward shaping:
+        reward = not_fall_bonus + truncked_walk_velocity - ctrl_cost
+    This task is not solved yet. But examples/tain_icub_sac.py can help the agent
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/icub_walk.py]: Nice catch. Forgot updating the comments.

File: python/social_bot/envs/icub_walk.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+            sub_seteps (int): take how many simulator substeps during one gym step
+        """
+        super(ICubWalk, self).__init__(port=port)
+        self._sub_seteps = sub_seteps
+        self._obs_stack = obs_stack
+        self._world = gazebo.new_world_from_file(
+            os.path.join(social_bot.get_world_dir(), "icub.world"))
+        self._agent = self._world.get_agent('icub')
+        # to avoid different parallel simulation has the same randomness
+        random.seed(port)
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/icub_walk.py]: will removed it

File: python/social_bot/envs/grocery_ground.py [hunk 12]
```diff
@@ -334,9 +343,17 @@ def step(self, action):
                           str(self._cum_reward))
         return obs, teacher_action.reward, teacher_action.done, {}
 
-    def _random_insert_objects(self):
-        for obj_id in range(len(self._object_list)):
-            model_name = self._object_list[obj_id]
+    def _insert_agent_to_world_file(self, world_file, model):
+        content = world_file.read()
+        insert_pos = content.find("<!-- Static Objects -->")
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: sure

File: python/social_bot/envs/grocery_ground.py [hunk 13]
```diff
@@ -319,7 +328,7 @@ def step(self, action):
         controls = dict(zip(self._agent_joints, controls))
         teacher_action = self._teacher.teach(sentence)
         self._agent.take_action(controls)
-        self._world.step(10)
+        self._world.step(50)
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: 10Hz update rate seems enough. I'll change this to 100ms.

File: pygazebo/pygazebo.cc [hunk 14]
```diff
@@ -440,11 +462,22 @@ PYBIND11_MODULE(pygazebo, m) {
...
+        &NewWorldFromString,
+        "Create a world from sdf string",
+        py::arg("std_string"));
+
   m.def("new_world_from_file",
         &NewWorldFromFile,
         "Create a world from .world file",
         py::arg("world_file"));
 
+  m.def("new_world_from_file_with_agent",
```
[the remark this answers is not in the document]
Comment (Jialn) [reply pygazebo/pygazebo.cc]: This is more stable than inserting the agent in python dynamically after the creation of the world, and removes the needs of 'time.sleep()' after the insertion.
Although we are inserting agent to world file for grocery env, I think this function could be useful latter.

File: python/social_bot/envs/grocery_ground.py [hunk 15]
```diff
@@ -255,10 +265,9 @@ def reset(self):
         self._collision_cnt = 0
         self._cum_reward = 0.0
         self._steps_in_this_episode = 0
-        self._agent.reset()
-        self._random_move_objects()
-        self._world.step(50)
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: Added

File: python/social_bot/envs/grocery_ground.py [hunk 16]
```diff
@@ -352,22 +369,23 @@ def _random_move_objects(self, random_range=10.0):
             loc = (obj_pos_list[obj_id][0], obj_pos_list[obj_id][1], 0)
             pose = (np.array(loc), (0, 0, 0))
             self._world.get_model(model_name).set_pose(pose)
-            self._world.step(5)
```
[the remark this answers is not in the document]
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: Yes

**Pull request merged**